// @typed/timer/cjs/Timeline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Timeline = void 0;
const lambda_1 = require("@typed/lambda");
const list_1 = require("@typed/list");
/**
 * Timeline is responsible for storing tasks at a given time and
 * returns what tasks are ready at the current time.
 */
class Timeline {
    constructor() {
        this.tasks = new Map();
        this.addTask = (time, f) => {
            const tasksAtTime = this.tasks.get(time) || [];
            tasksAtTime.push(f);
            this.tasks.set(time, tasksAtTime);
        };
        this.removeTask = (time, f) => {
            const tasksAtTime = this.tasks.get(time);
            if (!tasksAtTime) {
                return;
            }
            const index = tasksAtTime.findIndex((x) => x === f);
            if (index > -1) {
                tasksAtTime.splice(index, 1);
            }
            if (tasksAtTime.length === 0) {
                this.tasks.delete(time);
            }
        };
        this.readyTasks = (currentTime) => {
            const times = Array.from(this.tasks.keys());
            const timesToRun = times.filter((x) => x <= currentTime).sort(list_1.ascend(lambda_1.id));
            return list_1.chain(this.getAndDelete, timesToRun);
        };
        this.getAndDelete = (time) => {
            const tasks = this.tasks.get(time) || [];
            this.tasks.delete(time);
            return tasks;
        };
    }
}
exports.Timeline = Timeline;
