// @typed/disposable/esm/Disposable.js
import { noOp } from '@typed/lambda';
/**
 * Cleanup a disposable
 * @param disposable :: Disposable
 */
export const dispose = (disposable) => disposable.dispose();
export var Disposable;
(function (Disposable) {
    /**
     * Empty Disposable
     */
    Disposable.None = { dispose: noOp };
    /**
     * Create a disposable that is lazily created
     */
    Disposable.lazy = () => {
        let isDisposed = false;
        const disposables = new Set();
        const removeDisposable = (disposable) => disposables.delete(disposable);
        return {
            get disposed() {
                return isDisposed;
            },
            addDisposable(disposable) {
                if (disposable === Disposable.None) {
                    return disposable;
                }
                if (isDisposed) {
                    disposable.dispose();
                    return Disposable.None;
                }
                const dispose = () => removeDisposable(disposable);
                disposables.add(onDisposed(dispose, disposable));
                return {
                    dispose,
                };
            },
            dispose() {
                if (isDisposed) {
                    return;
                }
                isDisposed = true;
                disposables.forEach(dispose);
                disposables.clear();
            },
        };
    };
})(Disposable || (Disposable = {}));
// Useful when you have a non-cancellable (like promises) async process you want to be able to short circuit
export const withIsDisposed = (fn) => {
    let disposed = false;
    const isDisposed = () => disposed;
    fn(isDisposed);
    return {
        dispose: () => {
            disposed = true;
        },
    };
};
export function onDisposed(fn, disposable) {
    return {
        dispose: () => {
            try {
                disposable.dispose();
                fn();
            }
            catch (error) {
                fn(error);
            }
        },
    };
}
