// @mikro-orm/core/hydration/Hydrator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Hydrator = void 0;
const __1 = require("..");
class Hydrator {
    constructor(factory, em) {
        this.factory = factory;
        this.em = em;
    }
    /**
     * Hydrates the whole entity. This process handles custom type conversions, creating missing Collection instances,
     * mapping FKs to entity instances, as well as merging those entities.
     */
    hydrate(entity, meta, data, newEntity, convertCustomTypes) {
        const props = this.getProperties(meta, entity);
        for (const prop of props) {
            this.hydrateProperty(entity, prop, data, newEntity, convertCustomTypes);
        }
    }
    /**
     * Hydrates primary keys only
     */
    hydrateReference(entity, meta, data, convertCustomTypes) {
        const props = this.getProperties(meta, entity).filter(prop => prop.primary);
        for (const prop of props) {
            this.hydrateProperty(entity, prop, data, false, convertCustomTypes);
        }
    }
    getProperties(meta, entity) {
        const metadata = this.em.getMetadata();
        const root = __1.Utils.getRootEntity(metadata, meta);
        if (root.discriminatorColumn) {
            meta = metadata.find(entity.constructor.name);
        }
        return Object.values(meta.properties).filter(prop => {
            return !prop.inherited && root.discriminatorColumn !== prop.name && !prop.embedded;
        });
    }
}
exports.Hydrator = Hydrator;
